docs/Using-SRIOV-and-Clear-Containers.md

```diff
@@ -0,0 +1,218 @@
+virtual functions (VF) which enables direct pass through to virtual machines or
+containers.  For Clear Containers, we have enabled a CNM plugin and necessary
+changes to facilitate making use of SRIO-V for network based devices.
+
+## Install the SRIO-V Docker\* plugin
+
+To create a network with VFs associated with it which can be passed to
+Clear Containers, install a SRIOV Docker plugin. The created network will be
+based on a physical function (PF) device, and can create 'n' containers, where
+'n' is the number of VFs associated with the PF
+
+To install the plugin, you can follow the [plugin installation instructions](https://github.com/clearcontainers/sriov).
+
+
+## Host setup for SRIO-V
+
+You may need to rebuild your system's kernel in order to disable VFNOIOMMU in
+the kernel config and potentially add a PCI quirk for your NIC. If not, you can
+move to next section. The below section describes how to assess if
+changes are needed, and how to make these changes.
+
+### IOMMU Groups and PCIe Access Control Services
+Taking a look at how the iommu groups are setup on your host system can help
+provide information on whether or not your NIC is setup appropriately with
+respect to PCIe Access Control Services. More specifically, if the PCI bridge is
+within the same iommu-group as your NIC, then this is an indication that either
+your device doesn't support ACS or that it doesn't share this capability
+appropriately by default.
+
+For example, when you run the following command, if all is setup properly, you
+should have the PCI for each ACS enabled NIC port in its own iommu_group. If
+you don't see any output when running the following, then you likely need to
+update your host's kernel config to disable VFNOIOMMU.
+
+```
+find /sys/kernel/iommu_groups/ -type l
+```
+
+For more details, see http://vfio.blogspot.com/2014/08/iommu-groups-inside-and-out.html
+
+### Updating the Host Kernel
+
+These directions are provided as an example, and are based on Ubuntu 16.04. The
+user will need to do equivalent for their distribution.
+
+1.  Grab kernel sources
+
+```
+sudo apt-get install linux-source-4.10.0
+sudo apt-get install linux-headers-4.10.0
+cd /usr/src/linux-source-4.10.0/
+sudo tar -xvf linux-source-4.10.0.tar.bz2
+cd linux-source-4.10.0
+sudo apt-get install libssl-dev
+```
+
+2. Check the config and update if necessary
+
+```
+sudo cp /boot/config-4.8.0-36-generic .config
+sudo make olddefconfig #and verify resulting .config does not have NOIOMMU set; ie: # CONFIG_VFIO_NOIOMMU is not set
+```
+
+3.  If necessary, add PCI Quirk for SRIOV NIC
+
+Now, depending on how your NIC describes its ACS capabilities, you may need to add
+a quirk in order to indicate that the given NIC does properly support ACS.  An
+example is given below, but your mileage will vary (at the very least, check your
+PCI-ID).
+
+```
+modify drivers/pci/quirks.c:
+line 4118:
+static const u16 pci_quirk_intel_pch_acs_ids[] = {
```

Comment:
We're suggesting users change the contents of their host kernel. This needs two things imho:

- A warning explaining the risks of doing this (YMMV, "you get to keep the pieces...", etc).
- A suggestion that users also add a unique identifier to the `EXTRAVERSION` variable so that when the kernel is installed and running, running `uname -r` will show that this is a custom kernel.
- A warning explaining that if your system is configured for automatic distribution updates, you may lose this extra functionality the next time you run `dnf update`, `apt-get update`, etc.

Reply:
Archana, on your system did you need to make this change?  I think if this is a formal documentation / how-to, perhaps we shouldn't suggest this.  This'll make it look like the chipset supports true IOMMU, which may not be the case if you hack this driver...  perhaps we should just skip this, as it shouldn't be a normal occurrence for users...

Reply:
@amshinde ^

Reply:
@egernst No, I didnt need to do this, I was testing mainly on the test machine we have set up, where the 2 NIC ports were in separate IOMMU groups. I assumed you had to do this for your setup.
This is not formal documentation, but the kernel maintainer suggests  the approach to apply the patch here:
https://lkml.org/lkml/2013/5/30/513
 to quirk specific devices which enforce isolation but not provide an ACS capability. Note the vendor needs to confirm that the device supports isolation. I think I am going to take this out completely and instead add a link to to the pci-override patch , stating that this should be done only in case the device supports isolation.

Reply:
@amshinde - ok.  I think this should be rare, and only needed if they have an out of date kernel.  The quirk should be upstream already if their device doesn't advertise ACS correctly but does indeed support hardware isolation.  Worse case is they'd be marking it as supporting IOMMU when it indeed does not.  The link sounds best way to describe this rare case.

Reply:
@egernst Changes pushed for this.